[Swap][Tests][KEP-2400] Add swap serial stress tests, improve NodeConformance tests and adapt NoSwap behavior

iholder101 opened the pull request:
#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

Fixes https://github.com/kubernetes/kubernetes/issues/120798

#### What this PR does / why we need it:
This PR does the following:
1. Havily refactors and cleans swap node-e2e tests.
2. Adapt swap NodeConformance test to https://github.com/kubernetes/kubernetes/pull/122745 which drops `UnlimitedSwap` and introduces `NoSwap` behavior.
3. Adds serial tests for swap.

To expand on (3), two serial tests are now added which validate the following scenarios:
1. `"should be able over-commit the node memory"` - a pod is deployed which stresses memory. The test expects that once the pod uses all of the node's memory some of its memory would be swapped away.
2. `"should be able to use more memory than memory limits"` - a pod with memory limits stresses memory. Eventually it goes beyond its memory limits and instead of being OOM killed some of its memory is swapped away keeping it alive.

#### Special notes for your reviewer:
The original version of this PR was here: https://github.com/kubernetes/kubernetes/pull/120430

Due to unfortunate situations I had to be out of work for a while, therefore my PR was closed and continued here: https://github.com/kubernetes/kubernetes/pull/122175

Fortunately I was able to come back to work and therefore the PR is now continued here.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md
```

iholder101 committed:
Add NodeSwap as a node feature in nodefeature.go

Also, Remove wrong documentation about
tempSetCurrentKubeletConfig() returning bool

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Refactor: helper functions and quantity improvement

- Add getSleepingPod() helper function
- Refactor: quantity functions to return resource.quantity instead of int64
- Improve helper functions for memory capacity

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Move current test under its own NodeConformance context

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Remove cgroup v1 support for swap tests

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Fix swap feature gate check by introduting IsFeatureGateEnabled()

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
PR is not ready to review yet.

/uncc @dchen1107 @andrewsykim

kannon92 commented:
/cc

when it is ready

iholder101 commented:
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 committed:
Add serial tests

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-fedora-serial

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
/kind feature

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

kannon92 commented:
We have crio ci failures in upstream.

https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv1-node-e2e-features
https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv1-node-e2e-conformance
https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv2-node-e2e-conformance

I think you are impacted by this.

iholder101 committed:
Update node conformance to use NoSwap

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

kannon92 commented on test/e2e_node/swap_test.go:
```diff
@@ -38,49 +44,215 @@ import (
-		ginkgo.Entry("QOS Guaranteed", v1.PodQOSGuaranteed, false),
-	)
+			switch swapBehavior := getSwapBehavior(); swapBehavior {
+			case types.LimitedSwap:
```
I think there is no way this code will be hit in conformance test suite. We have to test the default path because we can't change the kubelet config in non-serial tests.

iholder101 replied:
> I think there is no way this code will be hit in conformance test suite.

Yes, this is true.
My reasoning behind this is that we might change defaults / what node conformance is using one day, then it would be less work. IOW, the test doesn't care which swap behavior you choose to run, it just makes sure that the right behavior occurs.

If you think it's better, we can instead expect that the behavior is indeed `NoSwap` and fail otherwise.

> we can't change the kubelet config in non-serial tests.

Right, and we don't :)

kannon92 replied:
> If you think it's better, we can instead expect that the behavior is indeed NoSwap and fail otherwise.

I would go this route. Less code is better in this scenario. I really don't see us changing the defaults and if we do I would rather have the tests fail and people update them.

iholder101 replied:
Done :+1:

iholder101 committed:
Expect NoSwap on NodeConformance, test LimitedSwap only in serial tests

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Refactor: Better pod image, read cgroup file from container

- Improve cgroup file read: execute from container instead of host
- Clean unused variables/functions
- BeTrue/BeFalse -> BeTrueBecause/BeFalseBecause
- Use agnhost instread of stress image
- Improve description and fix typo

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial
/test pull-kubernetes-node-e2e-containerd

iholder101 commented:
/reopen

iholder101 commented:
/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial
/test pull-kubernetes-node-e2e-containerd

iholder101 commented:
/retest

iholder101 commented:
@kannon92

NodeConformance tests have passed on `e2e-containerd`.
The serial lanes failed for seemingly unrelated reasons, but swap tests have passed there as well [[1](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767575322049384448)][[2](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-ubuntu-serial/1767575321990664192)].

The PR is ready to review :) PTAL

kannon92 commented:
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

Looks like these tests are timing out. Maybe your test is making them exceed the time limit allocated to them.

iholder101 commented:
> /test pull-kubernetes-node-swap-fedora-serial
> /test pull-kubernetes-node-swap-ubuntu-serial

(re-launching tests as I've pushed a minor change to satisfy the linter)
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

> Looks like these tests are timing out. Maybe your test is making them exceed the time limit allocated to them.

Cloud be.
I'll keep an eye on that.

iholder101 commented:
/test pull-kubernetes-node-e2e-containerd pull-kubernetes-dependencies

iholder101 commented:
Unrelated failure [[1](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767611214742949888)].

/test pull-kubernetes-node-swap-fedora-serial

kannon92 commented:
Looking at the serial logs, it looks like the agnhost container is being OOMed for fedora: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767625656390127616/artifacts/n1-standard-2-fedora-coreos-39-20240210-3-0-gcp-x86-64-436ed8ef/serial-1.log

I have been trying to chase down a really nasty bug with serial jobs that you seem to be hitting with ubuntu. 

https://github.com/kubernetes/kubernetes/issues/123589

Looking at those serial logs, I am seeing that we are hitting a OOM also for dd.

iholder101 commented:
> Looking at the serial logs, it looks like the agnhost container is being OOMed for fedora: [storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767625656390127616/artifacts/n1-standard-2-fedora-coreos-39-20240210-3-0-gcp-x86-64-436ed8ef/serial-1.log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767625656390127616/artifacts/n1-standard-2-fedora-coreos-39-20240210-3-0-gcp-x86-64-436ed8ef/serial-1.log)
> 
> I have been trying to chase down a really nasty bug with serial jobs that you seem to be hitting with ubuntu.
> 
> #123589
> 
> Looking at those serial logs, I am seeing that we are hitting a OOM also for dd.

So I've been tweaking things a bit, and all of the tests passed on a GCE instance (identical to the one we use on CI) for multiple times. Let's see if CI is on our side.

/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
Swap tests passed on both lanes, although they have failed. [[1](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767858931914248192)][[2](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-ubuntu-serial/1767858931855527936)].

@kannon92 do you mind help me figure out the flakiness of these lanes before we add the tests presented in this PR? Since the tests passed for 20 times in a row on an identical GCE instance, I wonder whether the lane is flaky by nature or if the newly introduced tests actually hurt its stability.

/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

kannon92 commented:
/test

kannon92 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

kannon92 commented on test/e2e/nodefeature/nodefeature.go:
```diff
@@ -100,6 +100,10 @@ var (
 
+	// Added to test Swap Feature
+	// This label should be used when testing KEP-2400 (Node Swap Support)
+	Swap = framework.WithNodeFeature(framework.ValidNodeFeatures.Add("NodeSwap"))
```
```suggestion
	Swap = framework.WithNodeFeature(framework.ValidNodeFeatures.Add("SwapConformance"))
```

I think.

iholder101 replied:
Not exactly.
The label you're pointing to here is for the NodeSwap feature gate. Instead I've added a `WithSwapConformance()` function to add the swap conformance label. Look at the [latest commit](https://github.com/kubernetes/kubernetes/pull/123557/commits/8a9a9617293551bddee0d4268fd7d0a71cedc1f5) for the implementation.

kannon92 replied:
Thats not true. This is just the label that we use for FOCUS.

iholder101 replied:
Right. That's what I meant. Let me rephrase.

We use `"NodeSwap"` at the upper `SIGDescribe`, meaning that this label appears for every swap test.
The new `"SwapConformance"` label introduced in this PR is applied only for the swap conformance tests.

This way:
- A job that focuses on NodeSwap runs all swap tests.
- A job that focuses on Serial will NOT run swap conformance.
- A job that focuses on SwapConformance will run swap conformance tests only.

Does this make sense? For more info look at the latest commit.

kannon92 replied:
So does "NodeSwap runs all swap tests" means all swap tests without SwapConformance?

iholder101 replied:
> So does "NodeSwap runs all swap tests" means all swap tests without SwapConformance?

No. Every swap test has the NodeSwap label. It makes sense, as if a job focuses only on NodeSwap it would be expected to run all swap related tests.

* All swap test has a NodeSwap label.
  * All swap node-conformance tests **also** have the NodeConformance label.
  * All swap swap-conformance tests **also** have the SwapConformance label.

Makes sense?

Please look at the code here which describes this:
https://github.com/iholder101/kubernetes/blob/635f936fcc3ffe902ef228defef4cfec367b2fbc/test/e2e_node/swap_test.go#L58-L85

iholder101 replied:
In short:
NodeConformance job would run only non-swap-conformance tests.
SwapConformance job would run only swap-conformance tests.

kannon92 replied:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-testing/3041-node-conformance-and-features could be a good read for you. 

NodeConformance is meant to be a test that can be run in all environments and not tied to any feature. SwapConformance seems to be meaningless and it is really a feature test of swap.

I'm going to tag @pohly and see what we think about `SwapConformance`.

iholder101 replied:
I just think that a test focuses on NodeSwap should run both node-conformance and non-node-conformance tests if they are related to the NodeSwap feature.

WDYT about replacing `SwapConformance` with something like `SerialSwap`?

In any case, if there's a strong push-back, I can add the NodeSwap label to serial tests only.

kannon92 replied:
I think having a swap enabled node run the node-conformance tests covers that case. So you could remove all swap related tests from the serial and non-serial swap jobs.

And in your "SwapConformance" jobs, you can just run all the tests that involve the swap feature. 

And then all the tests of the feature can be run on the serial node. You can test both NoSwap and LimitedSwap on serial because you can change the config.

I don't see a reason to add "SerialSwap". Serial is mostly just a implementation detail for the feature tests because we need to change from the default.

iholder101 replied:
@kannon92 
Alright, so let's summarize your suggestion to make sure we're on the same page here.

What you suggest is to remove `SwapConformance` from here:
https://github.com/iholder101/kubernetes/blob/635f936fcc3ffe902ef228defef4cfec367b2fbc/test/e2e_node/swap_test.go#L85

Then the swap-conformance jobs would focus on `NodeSwap` (i.e. `nodefeature.Swap`)?

This way node-conformance would run the non-stress swap tests
And swap-conformance would run ALL swap-related tests?

I'm fine with this approach

kannon92 replied:
Ideally, it would be nice to drop SwapConformance label entirely and just have all swap tests under NodeSwap. I think you titled the test-infra job for SwapConformance to target SwapConformance so we may need to change that first.

iholder101 replied:
@kannon92 Sure. Here's the PR: https://github.com/kubernetes/test-infra/pull/32276.

What are your thoughts on https://github.com/kubernetes/kubernetes/pull/123557#discussion_r1530313746?

kannon92 replied:
> BTW, I'm not sure what value we gain from running pull-kubernetes-node-swap-fedora-serial and pull-kubernetes-node-swap-ubuntu-serial, since they use NoSwap.

I think the goal was to make sure that enabling swap wouldn't break the conformance tests. I guess that since we defaulted from UnlimitedSwap to NoSwap we are only testing if swap will impact any of the tests.
> Isn't it (almost) similar to running the tests without swap enabled?
Don't we want to use LimitedSwap for these jobs?

We could but its not the default and changing a default in a non serial test is not really allowed. Changing a config requires a restart of kubelet..

We could change the serial jobs to use LimitedSwap but not sure what value that would have.

I could see us dropping those lanes eventually but I say we should keep them in case.

iholder101 replied:
> > BTW, I'm not sure what value we gain from running pull-kubernetes-node-swap-fedora-serial and pull-kubernetes-node-swap-ubuntu-serial, since they use NoSwap.
> 
> I think the goal was to make sure that enabling swap wouldn't break the conformance tests. I guess that since we defaulted from UnlimitedSwap to NoSwap we are only testing if swap will impact any of the tests.
> 
> > Isn't it (almost) similar to running the tests without swap enabled?
> > Don't we want to use LimitedSwap for these jobs?
> 
> We could but its not the default and changing a default in a non serial test is not really allowed. Changing a config requires a restart of kubelet..

Right. The lanes I've mentioned are serial.

> We could change the serial jobs to use LimitedSwap but not sure what value that would have.
> 
> I could see us dropping those lanes eventually but I say we should keep them in case.

I just think that at this point we have certain confidence in the fact that swap enabled on the node doesn't break serial tests. Running these tests with LimitedSwap would give raise our confidence to a much higher level since we'd know that not only a node with swap enabled doesn't break stuff, but actually using swap with LimitedSwap keeps the node stable.

Just a thought though, not a must.

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
@kannon92 Looking good! :tada: 

/test pull-kubernetes-node-swap-fedora-serial

iholder101 commented:
The fact that `swap-fedora-serial` is failing even without running swap conformance tests shows it is probably flaky and that this PR is not the cause.

Trying one last time.

/test pull-kubernetes-node-swap-fedora-serial

iholder101 commented on test/e2e_node/swap_test.go:
```diff
@@ -38,49 +45,229 @@ import (
+						}
+
+						// Better to delete the stress pod ASAP to avoid node failures
+						_ = podClient.Delete(context.Background(), stressPod.Name, metav1.DeleteOptions{})
```
> I don't think we should call this more than once?
> Could we move it out of the eventually?

It is being called only once before the Eventually clause returns no err (nil).
I basically wanted to delete the pod ASAP to avoid compromising the node, but there's no reason to do so right after the eventually clause ends.

Thanks @kannon92! I agree, the code is much clearer this way.
Done.

> And can we assert there is no error?

Sure. Done.

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

iholder101 commented:
/retest

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

iholder101 commented:
@kannon92 could you please have another look now that https://github.com/kubernetes/test-infra/pull/32276 is merged and the swap conformance label is removed?

kannon92 commented:
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

kannon92 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial
/test pull-kubernetes-node-swap-conformance-ubuntu-serial

kannon92 commented:
tests look good now

/lgtm

kannon92 commented:
/triage accepted
/priority important-soon

Let's get this into 1.31 as soon as it opens.

iholder101 commented:
@mrunalp Would you mind taking a look?

kannon92 commented:
/assign @mrunalp @SergeyKanzhelev @dchen1107

SergeyKanzhelev commented:
/retest

iholder101 commented:
> /retest

Hey @SergeyKanzhelev!
Please note that the regular `pull-kubernetes-node-swap-*-serial` lanes are flaky.

We've introduced the new `pull-kubernetes-node-swap-conformance-*-serial` lanes, which are stable and run only swap-stress related tests, for that purpose. These tests pass continuously on this PR.

pohly commented:
Is someone working on fixing this flakiness?

Fixing flakes should have more priority than adding more tests or features.

iholder101 commented:
> Is someone working on fixing this flakiness?
> 
> Fixing flakes should have more priority than adding more tests or features.

While I agree it's important, I think it's out-of-scope for this PR.

This PR does not affect the tests that these lanes are running, but rather adds new serial stress tests to run under `pull-kubernetes-node-swap-conformance-*-serial` lanes.

kannon92 commented:
@pohly We are working on fixing flakyness in #sig-node-ci but bandwidth is limited. I don't think we can block all features in sig-node until we have a green CI. 

Fixing flakes is complicated because each test can have a different reason. For example, we have a flake in our eviction tests (for at least two years) and we spent at least a few months working on fixing it. It turns out that there was a limitation in our cadvisor stats code which was reporiting wrong processes. We debated about fixing at the cadvisor level or at the stats level.

haircommander commented on test/e2e_node/swap_test.go:
```diff
@@ -38,49 +45,231 @@ import (
+	noLimits *resource.Quantity = nil
+)
+
+var _ = SIGDescribe("Swap", "[LinuxOnly]", nodefeature.Swap, func() {
```
nevermind, that's what the variable is

haircommander commented:
/lgtm

serial tests are flaking but the new test passed, thanks @iholder101

rphillips commented:
lgtm as well

SergeyKanzhelev commented on test/e2e_node/swap_test.go:
```diff
@@ -197,57 +406,137 @@ func expectLimitedSwap(f *framework.Framework, pod *v1.Pod, expectedSwapLimit in
+
+// Sets memory request or limit can be null, then it's dismissed.
+// Sets the same value for all containers.
+func setPodMmoryResources(pod *v1.Pod, memoryRequest, memoryLimit *resource.Quantity) {
```
is misspelling of the name intentional?

iholder101 replied:
It isn't, thanks for catching that! fixed

SergeyKanzhelev commented on test/e2e_node/swap_test.go:
```diff
@@ -38,49 +45,231 @@ import (
+					return pod
+				}
+
+				ginkgo.It("should be able over-commit the node memory", func() {
```
minor: over-commit sounds like we can request more than node allows. I'd call it should be able to use more than the node memory capacity

iholder101 replied:
Done!

SergeyKanzhelev commented on test/e2e_node/swap_test.go:
```diff
@@ -38,49 +45,231 @@ import (
+			)
+		})
+
+		// These tests assume the following, and will fail otherwise:
```
thank you for those. What swappiness value did you use in your configs? is it something we need to take into consideration. I understand that kernel may not be super efficient in swapping some times. So depending on how fast we write memory, test may start flaking.

The soft memory limit in cgroupv2 is something that I think may improve swap reliability.

iholder101 replied:
>thank you for those. What swappiness value did you use in your configs? is it something we need to take into consideration.

Thanks for your review Sergey!

I've spent a lot of time writing these tests in a way that they will be as non-flaky and stable as possible, but at the same time will verify that every expectation from the NodeSwap feature is met. I've followed a few principles in order to achieve it:
* **The system is not configured in any special way**: The tests should work with default OS configurations to prove that NodeSwap is not dependent on any specific configuration like swappiness.
* **Tests end once the most basic functionality is verified**: For example, when ensuring that a container can exceed memory limits, the container (slowly but surely) stresses memory and once `memory.current > limits` and `memory.swap.current > 0` the test passes.

> So depending on how fast we write memory, test may start flaking.

Bear in mind that the stress is being made at a low pace to give the kernel enough time to reclaim. It is proven to be very stable until now, but we can always make it even slower if we'll see future flakiness.

I think that swappiness shouldn't matter too much here as well. Essentially it is a way to tell the kernel to prefer reclaiming file-backed memory vs anon memory, but in any case after enough stress the kernel would have to start swapping.

> The soft memory limit in cgroupv2 is something that I think may improve swap reliability.

I think you're referring to `memory.swap.high`.
This is indeed an interesting topic which I think about a lot. I'm not sure if it's in scope for this PR, but I would love discussing it offline if you have the time/interest.

iholder101 commented:
@SergeyKanzhelev @mrunalp
Can you PTAL?

iholder101 commented:
/test pull-kubernetes-node-e2e-containerd

iholder101 commented:
/test pull-kubernetes-node-swap-conformance-fedora-serial

iholder101 commented:
/test pull-kubernetes-integration

iholder101 commented:
@dchen1107 @kannon92 can you PTAL?

haircommander commented:
/lgtm

pohly commented on test/e2e/framework/skipper/skipper.go:
```diff
@@ -67,6 +66,13 @@ func InitFeatureGates(defaults featuregate.FeatureGate, overrides map[string]boo
 	return nil
 }
 
+func IsFeatureGateEnabled(feature featuregate.Feature) bool {
```
Can you added documentation? Feature gate testing is one of those areas that have been misunderstood by test authors in the past.

I'll re-add LGTM and approve when done.

/lgtm cancel

iholder101 replied:
Thanks @pohly!
An important comment indeed. I'm also a victim of misunderstanding how it works :)

I've added the following documentation, let me know what you think:
> // IsFeatureGateEnabled can be used during e2e tests to figure out if a certain feature gate is enabled.
> // This function is dependent on InitFeatureGates under the hood. Therefore, the test must be called with a
> // --feature-gates parameter.

iholder101 commented:
/retest
